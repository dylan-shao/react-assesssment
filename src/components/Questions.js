import React from 'react';
import PropTypes from 'prop-types';

const Questions = (props) => {
  const { questionId, points, content, choices, clickHandler } = props;
  const bindClick = (id, select) => (clickHandler.bind(null, [id, select]));
  const isChoosen = (cid) => {
    if (points[questionId] && points[questionId].cid === cid) {
      return true;
    }
    return false;
  };
  const renderChoices = choice => (
    <li key={choice.cid}>
      <button
        id={`choice-${choice.cid}`}
        className={isChoosen(choice.cid) ? 'button-choosen button-item' : 'button-item'}
        onClick={bindClick(questionId, choice)}
      >{choice.content}
      </button>
    </li>
  );

  if (content && choices) {
    return (
      <div className="questions clear-both">
        <h2 className="question-content">{content}</h2>
        <div className="choices">
          <ul>{choices.map(renderChoices)}</ul>
        </div>
      </div>
    );
  }
  return (
    <div>Loading</div>
  );
};

const { func, number, string, arrayOf, objectOf, object } = PropTypes;

Questions.propTypes = {
  clickHandler: func.isRequired,
  questionId: number.isRequired,
  content: string.isRequired,
  choices: arrayOf(object).isRequired,
  points: objectOf(object).isRequired,
};

export default Questions;
