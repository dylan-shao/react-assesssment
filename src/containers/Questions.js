import React, { Component } from 'react';
import Choices from '../components/Choices';
import { saveQuestionPoint } from '../store/actions';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';


class Questions extends Component {

	shouldComponentUpdate(nextProps, nextStates){
		return this.props.questionId !== nextProps.questionId;
	}

	componentDidUpdate() {
		const el = document.getElementById(this.props.focusId);
		el && el.focus();
	}

	clickHandler = (obj) => {
		this.props.saveQuestionPoint(obj);
	}

	render() {
		const { questionId, content, choices } = this.props;
		if(content && choices){
			return (
				<div className="questions clear-both">
					<h2 className="question-content">{content}</h2>
			        <Choices questionId={questionId} choices={choices} clickHandler={this.clickHandler}/>
				</div>
			);
		} else {
			return (
				<div>Loading</div>
			);
		}
		
	}
}

function mapDispatchToProps(dispatch) {
	return bindActionCreators({saveQuestionPoint}, dispatch);
}
export default connect(null, mapDispatchToProps)(Questions);