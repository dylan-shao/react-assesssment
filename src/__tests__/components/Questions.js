import React from 'react';
import { expect } from 'chai';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import renderer from 'react-test-renderer';
import Questions from '../../components/Questions';

const ee = global.expect;
describe('<Question />', () => {
  let defaultProps;
  beforeEach(function(){
	defaultProps = {
		clickHandler: sinon.spy(),
		questionId: 0,
		content: 'content',
		choices: [{cid: 1, content: 'choice1'}, 
  	      {cid: 2, content: 'choice2'}, 
  	      {cid: 3, content: 'choice3'}],
		points: {}
  	};
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(<Questions {...defaultProps}/>)
      .toJSON();
    ee(tree).toMatchSnapshot();
  });
  test('should render Loading if content and choices deoes not exist', function() {
  	defaultProps.content = '';
  	const wrapper = shallow(<Questions {...defaultProps}/>);
  	expect(wrapper.text()).to.equal('Loading');
  });

  test('should contain h2 with content', () => {
  	const wrapper = shallow(<Questions {...defaultProps}/>);
  	expect(wrapper.find('h2').text()).to.equal(defaultProps.content);
  });

  test('should render a list with button based on choices', () => {
  	const wrapper = shallow(<Questions {...defaultProps}/>);
  	expect(wrapper.find('li button').length).to.equal(defaultProps.choices.length);
  });

  test('should trigger clickHandler when click the choices', () => {
  	const wrapper = shallow(<Questions {...defaultProps}/>);
  	wrapper.find('li button').first().simulate('click');
  	expect(defaultProps.clickHandler.callCount).to.equal(1);
  });
});