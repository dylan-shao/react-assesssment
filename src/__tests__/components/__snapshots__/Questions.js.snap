// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Question /> renders correctly 1`] = `
<div
  className="questions clear-both"
>
  <h2
    className="question-content"
  >
    content
  </h2>
  <div
    className="choices"
  >
    <ul>
      <li>
        <button
          className="button-item"
          onClick={[Function]}
        >
          choice1
        </button>
      </li>
      <li>
        <button
          className="button-item"
          onClick={[Function]}
        >
          choice2
        </button>
      </li>
      <li>
        <button
          className="button-item"
          onClick={[Function]}
        >
          choice3
        </button>
      </li>
    </ul>
  </div>
</div>
`;
